selfhost/typechecker.jakt

```diff
@@ -6194,26 +6230,41 @@ struct Typechecker {
         let maybe_function_id = .find_function_in_scope(parent_scope_id: current_scope_id, function_name: call.name)
         if maybe_function_id.has_value() {
             let function_id = maybe_function_id!
-            if not must_be_enum_constructor or .get_function(function_id).type is ImplicitEnumConstructor {
+            let func = .get_function(function_id)
+            if not must_be_enum_constructor or func.type is ImplicitEnumConstructor {
                 return function_id
             }
         }
 
         if must_be_enum_constructor {
             .error(format("No such enum constructor ‘{}’", call.name), span)
-            return callee
+            return None
         }
 
         // 2. Look for a struct, class or enum constructor with this name.
         let maybe_struct_id = .find_struct_in_scope(scope_id: current_scope_id, name: call.name)
         if maybe_struct_id.has_value() {
             let struct_id = maybe_struct_id!
             let structure = .get_struct(struct_id)
-            let maybe_function_id = .find_function_in_scope(parent_scope_id: structure.scope_id, function_name: call.name)
-            if maybe_function_id.has_value() {
-                return maybe_function_id!
+            let maybe_explicit_constructor_id = .find_function_in_scope(parent_scope_id: structure.scope_id, function_name: "create")
+            let maybe_implicit_constructor_id = .find_function_in_scope(parent_scope_id: structure.scope_id, function_name: call.name)
+            if maybe_explicit_constructor_id.has_value() {
+                let explicit_constructor = .get_function(maybe_explicit_constructor_id!)
+                if explicit_constructor.type is ExplicitConstructor {
+                    if .scope_can_access(accessor: scope_id, accessee: explicit_constructor.block.scope_id) {
+                        // fixme: can't we assert this?
```

Comment:
`maybe_implicit_constructor_id!` has a built-in `VERIFY()`, so can we just drop the `has_value()` check to begin with?

Reply:
I am not sure I understand. The `!` always come after a has value check afaict. there is the question wether we can assert the existence of an explicit constructor, but I am not 100% sure.

Reply:
I.e. in theory an implicit constructor should always exist, but I don't quite understand how external constructors work, so I didnt dare yet :).